CI is failing with:
```
[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   ServletITCase>ServletTestCase.testWebjars:73 1 expectation failed.
Expected status code <200> but was <404>.

[ERROR]   ServletITCase>ServletTestCase.testWelcomeFile:51 1 expectation failed.
Response body doesn't match expectation.
Expected: a string containing "A HTML page"
  Actual: RESTEASY003210: Could not find resource for full path: http://localhost:8081/

[ERROR]   SwaggerUIITCase>SwaggerUITestCase.testSwaggerUi:32 1 expectation failed.
Response body doesn't match expectation.
Expected: a string containing "#swagger-ui"
  Actual: RESTEASY003210: Could not find resource for full path: http://localhost:8081/swagger-ui/

[ERROR] Errors: 
[ERROR]   OpenApiITCase>OpenApiTestCase.testOpenAPIJSON:49 » IO Server returned HTTP res...
```

--- a/core/deployment/src/main/java/io/quarkus/deployment/util/FileUtil.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/util/FileUtil.java
@@ -1,6 +1,8 @@
 package io.quarkus.deployment.util;
 
+import java.io.ByteArrayOutputStream;
 import java.io.IOException;
+import java.io.InputStream;
 import java.nio.file.FileVisitResult;
 import java.nio.file.Files;
 import java.nio.file.Path;
@@ -35,5 +37,15 @@ public class FileUtil {
             }
 
         });
+    }
+
+    public static byte[] readFileContents(InputStream inputStream) throws IOException {
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        byte[] buf = new byte[10000];
+        int r;
+        while ((r = inputStream.read(buf)) > 0) {
+            out.write(buf, 0, r);
+        }
+        return out.toByteArray();
     }
 }

--- a/core/deployment/src/main/java/io/quarkus/runner/RuntimeClassLoader.java
+++ b/core/deployment/src/main/java/io/quarkus/runner/RuntimeClassLoader.java
@@ -429,7 +429,7 @@ public class RuntimeClassLoader extends ClassLoader implements ClassOutput, Tran
     private URL getQuarkusResource(String name) {
         byte[] data = resources.get(name);
         if (data != null) {
-            String path = "quarkus:" + name + "/";
+            String path = "quarkus:" + name;
 
             try {
                 URL url = new URL(null, path, new URLStreamHandler() {

--- a/extensions/smallrye-openapi/deployment/src/main/java/io/quarkus/smallrye/openapi/deployment/SmallRyeOpenApiProcessor.java
+++ b/extensions/smallrye-openapi/deployment/src/main/java/io/quarkus/smallrye/openapi/deployment/SmallRyeOpenApiProcessor.java
@@ -49,16 +49,15 @@ import io.quarkus.deployment.annotations.BuildStep;
 import io.quarkus.deployment.builditem.ApplicationArchivesBuildItem;
 import io.quarkus.deployment.builditem.CombinedIndexBuildItem;
 import io.quarkus.deployment.builditem.FeatureBuildItem;
-import io.quarkus.deployment.builditem.GeneratedResourceBuildItem;
 import io.quarkus.deployment.builditem.HotDeploymentWatchedFileBuildItem;
 import io.quarkus.deployment.builditem.substrate.ReflectiveClassBuildItem;
 import io.quarkus.deployment.builditem.substrate.ReflectiveHierarchyBuildItem;
-import io.quarkus.deployment.builditem.substrate.SubstrateResourceBuildItem;
 import io.quarkus.deployment.logging.LogCleanupFilterBuildItem;
 import io.quarkus.resteasy.deployment.ResteasyJaxrsConfigBuildItem;
 import io.quarkus.smallrye.openapi.common.deployment.SmallRyeOpenApiConfig;
 import io.quarkus.smallrye.openapi.runtime.OpenApiDocumentProducer;
 import io.quarkus.smallrye.openapi.runtime.OpenApiServlet;
+import io.quarkus.undertow.deployment.GeneratedWebResourceBuildItem;
 import io.quarkus.undertow.deployment.ServletBuildItem;
 import io.smallrye.openapi.api.OpenApiConfig;
 import io.smallrye.openapi.api.OpenApiConfigImpl;
@@ -198,8 +197,7 @@ public class SmallRyeOpenApiProcessor {
     public void build(ApplicationArchivesBuildItem archivesBuildItem,
             CombinedIndexBuildItem combinedIndexBuildItem, BuildProducer<FeatureBuildItem> feature,
             Optional<ResteasyJaxrsConfigBuildItem> resteasyJaxrsConfig,
-            BuildProducer<GeneratedResourceBuildItem> resourceBuildItemBuildProducer,
-            BuildProducer<SubstrateResourceBuildItem> substrateResourceBuildItemBuildProducer) throws Exception {
+            BuildProducer<GeneratedWebResourceBuildItem> resourceBuildItemBuildProducer) throws Exception {
         feature.produce(new FeatureBuildItem(FeatureBuildItem.SMALLRYE_OPENAPI));
         OpenAPI staticModel = generateStaticModel(archivesBuildItem);
         OpenAPI annotationModel;
@@ -210,10 +208,9 @@ public class SmallRyeOpenApiProcessor {
         }
         OpenApiDocument finalDocument = loadDocument(staticModel, annotationModel);
         for (OpenApiSerializer.Format format : OpenApiSerializer.Format.values()) {
-            String name = OpenApiServlet.BASE_NAME + format;
-            resourceBuildItemBuildProducer.produce(new GeneratedResourceBuildItem(name,
+            String name = OpenApiServlet.GENERATED_DOC_BASE + format;
+            resourceBuildItemBuildProducer.produce(new GeneratedWebResourceBuildItem(name,
                     OpenApiSerializer.serialize(finalDocument.get(), format).getBytes(StandardCharsets.UTF_8)));
-            substrateResourceBuildItemBuildProducer.produce(new SubstrateResourceBuildItem(name));
         }
     }
 

--- a/extensions/smallrye-openapi/runtime/src/main/java/io/quarkus/smallrye/openapi/runtime/OpenApiServlet.java
+++ b/extensions/smallrye-openapi/runtime/src/main/java/io/quarkus/smallrye/openapi/runtime/OpenApiServlet.java
@@ -36,7 +36,7 @@ public class OpenApiServlet extends HttpServlet {
 
     private static final String QUERY_PARAM_FORMAT = "format";
 
-    protected static final String GENERATED_DOC_BASE = "quarkus-generated-openapi-doc.";
+    public static final String GENERATED_DOC_BASE = "quarkus-generated-openapi-doc.";
     public static final String BASE_NAME = "META-INF/resources/" + GENERATED_DOC_BASE;
 
     @Override

--- a/extensions/swagger-ui/deployment/src/main/java/io/quarkus/swaggerui/deployment/SwaggerUiProcessor.java
+++ b/extensions/swagger-ui/deployment/src/main/java/io/quarkus/swaggerui/deployment/SwaggerUiProcessor.java
@@ -5,6 +5,7 @@ import static java.lang.String.format;
 import java.io.File;
 import java.io.IOException;
 import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
@@ -33,6 +34,7 @@ import io.quarkus.runtime.annotations.ConfigItem;
 import io.quarkus.runtime.annotations.ConfigRoot;
 import io.quarkus.smallrye.openapi.common.deployment.SmallRyeOpenApiConfig;
 import io.quarkus.swaggerui.runtime.SwaggerUiTemplate;
+import io.quarkus.undertow.deployment.GeneratedWebResourceBuildItem;
 import io.quarkus.undertow.deployment.ServletExtensionBuildItem;
 
 public class SwaggerUiProcessor {
@@ -41,6 +43,7 @@ public class SwaggerUiProcessor {
 
     private static final String SWAGGER_UI_WEBJAR_GROUP_ID = "org.webjars";
     private static final String SWAGGER_UI_WEBJAR_ARTIFACT_ID = "swagger-ui";
+    private static final String META_INF_RESOURCES = "META-INF/resources/";
     private static final String SWAGGER_UI_WEBJAR_PREFIX = "META-INF/resources/webjars/swagger-ui";
     private static final Pattern SWAGGER_UI_DEFAULT_API_URL_PATTERN = Pattern.compile("(.* url: \")(.*)(\",.*)",
             Pattern.DOTALL);
@@ -58,7 +61,7 @@ public class SwaggerUiProcessor {
 
     @BuildStep
     void feature(BuildProducer<FeatureBuildItem> feature) {
-        if (launch.getLaunchMode().isDevOrTest()) {
+        if (launch.getLaunchMode().isDevOrTest() || swaggerUiConfig.alwaysInclude) {
             feature.produce(new FeatureBuildItem(FeatureBuildItem.SWAGGER_UI));
         }
     }
@@ -68,7 +71,8 @@ public class SwaggerUiProcessor {
     public void registerSwaggerUiServletExtension(SwaggerUiTemplate template,
             BuildProducer<ServletExtensionBuildItem> servletExtension,
             BeanContainerBuildItem container,
-            LiveReloadBuildItem liveReloadBuildItem) {
+            BuildProducer<GeneratedWebResourceBuildItem> generatedResources,
+            LiveReloadBuildItem liveReloadBuildItem) throws Exception {
 
         if (launch.getLaunchMode().isDevOrTest()) {
             CachedSwaggerUI cached = liveReloadBuildItem.getContextObject(CachedSwaggerUI.class);
@@ -105,6 +109,30 @@ public class SwaggerUiProcessor {
                                     swaggerUiConfig.path,
                                     cached.cachedDirectory,
                                     container.getValue())));
+        } else if (swaggerUiConfig.alwaysInclude) {
+            ResolvedArtifact artifact = getSwaggerUiArtifact();
+            //we are including in a production artifact
+            //just stick the files in the generated output
+            //we could do this for dev mode as well but then we need to extract them every time
+            File artifactFile = artifact.getArtifactPath().toFile();
+            try (JarFile jarFile = new JarFile(artifactFile)) {
+                Enumeration<JarEntry> entries = jarFile.entries();
+                String versionedSwaggerUiWebjarPrefix = format("%s/%s/", SWAGGER_UI_WEBJAR_PREFIX, artifact.getVersion());
+                while (entries.hasMoreElements()) {
+                    JarEntry entry = entries.nextElement();
+                    if (entry.getName().startsWith(versionedSwaggerUiWebjarPrefix) && !entry.isDirectory()) {
+                        InputStream inputStream = jarFile.getInputStream(entry);
+                        String filename = entry.getName().replace(versionedSwaggerUiWebjarPrefix, "");
+                        byte[] content = FileUtil.readFileContents(inputStream);
+                        if (entry.getName().endsWith("index.html")) {
+                            content = updateApiUrl(new String(content, StandardCharsets.UTF_8))
+                                    .getBytes(StandardCharsets.UTF_8);
+                        }
+                        generatedResources
+                                .produce(new GeneratedWebResourceBuildItem(swaggerUiConfig.path + "/" + filename, content));
+                    }
+                }
+            }
         }
     }
 
@@ -130,13 +158,21 @@ public class SwaggerUiProcessor {
     }
 
     private void updateApiUrl(Path indexHtml) throws IOException {
-        String content = new String(Files.readAllBytes(indexHtml), "UTF-8");
-        Matcher uriMatcher = SWAGGER_UI_DEFAULT_API_URL_PATTERN.matcher(content);
+        String content = new String(Files.readAllBytes(indexHtml), StandardCharsets.UTF_8);
+        String result = updateApiUrl(content);
+        if (result != null) {
+            Files.write(indexHtml, result.getBytes(StandardCharsets.UTF_8));
+        }
+    }
+
+    public String updateApiUrl(String original) {
+
+        Matcher uriMatcher = SWAGGER_UI_DEFAULT_API_URL_PATTERN.matcher(original);
         if (uriMatcher.matches()) {
-            content = uriMatcher.replaceFirst("$1" + openapi.path + "$3");
-            Files.write(indexHtml, content.getBytes("UTF-8"));
+            return uriMatcher.replaceFirst("$1" + openapi.path + "$3");
         } else {
             log.warn("Unable to replace the default URL of Swagger UI");
+            return null;
         }
     }
 
@@ -147,6 +183,13 @@ public class SwaggerUiProcessor {
          */
         @ConfigItem(defaultValue = "/swagger-ui")
         String path;
+
+        /**
+         * If this should be included every time. By default this is only included when the application is running
+         * in dev mode.
+         */
+        @ConfigItem(defaultValue = "false")
+        boolean alwaysInclude;
     }
 
     private static final class CachedSwaggerUI implements Runnable {

--- a/extensions/undertow/deployment/src/main/java/io/quarkus/undertow/deployment/GeneratedWebResourceBuildItem.java
+++ b/extensions/undertow/deployment/src/main/java/io/quarkus/undertow/deployment/GeneratedWebResourceBuildItem.java
@@ -1,0 +1,42 @@
+/*
+ * Copyright 2018 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.quarkus.undertow.deployment;
+
+import io.quarkus.builder.item.MultiBuildItem;
+
+/**
+ * A generated static resource that will be served by the web container
+ *
+ * This is automatically registered in substrate
+ */
+public final class GeneratedWebResourceBuildItem extends MultiBuildItem {
+    final String name;
+    final byte[] classData;
+
+    public GeneratedWebResourceBuildItem(String name, byte[] classData) {
+        this.name = name;
+        this.classData = classData;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    public byte[] getClassData() {
+        return classData;
+    }
+}

--- a/extensions/undertow/deployment/src/main/java/io/quarkus/undertow/deployment/KnownPathsBuildItem.java
+++ b/extensions/undertow/deployment/src/main/java/io/quarkus/undertow/deployment/KnownPathsBuildItem.java
@@ -1,0 +1,16 @@
+package io.quarkus.undertow.deployment;
+
+import java.util.Set;
+
+import io.quarkus.builder.item.SimpleBuildItem;
+
+final class KnownPathsBuildItem extends SimpleBuildItem {
+
+    final Set<String> knownFiles;
+    final Set<String> knownDirectories;
+
+    KnownPathsBuildItem(Set<String> knownFiles, Set<String> knownDirectories) {
+        this.knownFiles = knownFiles;
+        this.knownDirectories = knownDirectories;
+    }
+}

--- a/extensions/undertow/deployment/src/main/java/io/quarkus/undertow/deployment/UndertowBuildStep.java
+++ b/extensions/undertow/deployment/src/main/java/io/quarkus/undertow/deployment/UndertowBuildStep.java
@@ -23,21 +23,17 @@ import static io.undertow.servlet.api.SecurityInfo.EmptyRoleSemantic.PERMIT;
 import static javax.servlet.DispatcherType.REQUEST;
 
 import java.io.IOException;
-import java.net.JarURLConnection;
 import java.net.URL;
-import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
-import java.util.Enumeration;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
 import java.util.function.Consumer;
-import java.util.jar.JarEntry;
 import java.util.stream.Collectors;
 
 import javax.annotation.security.DeclareRoles;
@@ -95,7 +91,6 @@ import io.quarkus.arc.deployment.AdditionalBeanBuildItem;
 import io.quarkus.arc.deployment.BeanContainerBuildItem;
 import io.quarkus.arc.deployment.ContextRegistrarBuildItem;
 import io.quarkus.arc.processor.ContextRegistrar;
-import io.quarkus.deployment.ApplicationArchive;
 import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.annotations.BuildStep;
 import io.quarkus.deployment.annotations.Record;
@@ -132,6 +127,7 @@ import io.undertow.servlet.api.ServletInfo;
 import io.undertow.servlet.api.ServletSecurityInfo;
 import io.undertow.servlet.handlers.DefaultServlet;
 
+//TODO: break this up, it is getting too big
 public class UndertowBuildStep {
 
     public static final DotName WEB_FILTER = DotName.createSimple(WebFilter.class.getName());
@@ -141,7 +137,6 @@ public class UndertowBuildStep {
     public static final DotName DECLARE_ROLES = DotName.createSimple(DeclareRoles.class.getName());
     public static final DotName MULTIPART_CONFIG = DotName.createSimple(MultipartConfig.class.getName());
     public static final DotName SERVLET_SECURITY = DotName.createSimple(ServletSecurity.class.getName());
-    protected static final String META_INF_RESOURCES = "META-INF/resources";
     protected static final String SERVLET_CONTAINER_INITIALIZER = "META-INF/services/javax.servlet.ServletContainerInitializer";
     protected static final DotName HANDLES_TYPES = DotName.createSimple(HandlesTypes.class.getName());
 
@@ -273,8 +268,7 @@ public class UndertowBuildStep {
 
     @Record(STATIC_INIT)
     @BuildStep()
-    public ServletDeploymentManagerBuildItem build(ApplicationArchivesBuildItem applicationArchivesBuildItem,
-            List<ServletBuildItem> servlets,
+    public ServletDeploymentManagerBuildItem build(List<ServletBuildItem> servlets,
             List<FilterBuildItem> filters,
             List<ListenerBuildItem> listeners,
             List<ServletInitParamBuildItem> initParams,
@@ -288,7 +282,7 @@ public class UndertowBuildStep {
             Consumer<ReflectiveClassBuildItem> reflectiveClasses,
             LaunchModeBuildItem launchMode,
             ShutdownContextBuildItem shutdownContext,
-            BuildProducer<SubstrateResourceBuildItem> substrateResourceBuildItemBuildProducer) throws Exception {
+            KnownPathsBuildItem knownPaths) throws Exception {
 
         ObjectSubstitutionBuildItem.Holder holder = new ObjectSubstitutionBuildItem.Holder(ServletSecurityInfo.class,
                 ServletSecurityInfoProxy.class, ServletSecurityInfoSubstitution.class);
@@ -296,55 +290,8 @@ public class UndertowBuildStep {
         reflectiveClasses.accept(new ReflectiveClassBuildItem(false, false, DefaultServlet.class.getName(),
                 "io.undertow.server.protocol.http.HttpRequestParser$$generated"));
 
-        //we need to check for web resources in order to get welcome files to work
-        //this kinda sucks
-        Set<String> knownFiles = new HashSet<>();
-        Set<String> knownDirectories = new HashSet<>();
-        for (ApplicationArchive i : applicationArchivesBuildItem.getAllApplicationArchives()) {
-            Path resource = i.getChildPath(META_INF_RESOURCES);
-            if (resource != null && Files.exists(resource)) {
-                Files.walk(resource).forEach(new Consumer<Path>() {
-                    @Override
-                    public void accept(Path path) {
-                        // Skip META-INF/resources entry
-                        if (resource.equals(path)) {
-                            return;
-                        }
-                        Path rel = resource.relativize(path);
-                        if (Files.isDirectory(rel)) {
-                            knownDirectories.add(rel.toString());
-                        } else {
-                            knownFiles.add(rel.toString());
-                        }
-                    }
-                });
-            }
-        }
-        Enumeration<URL> resources = getClass().getClassLoader().getResources(META_INF_RESOURCES);
-        while (resources.hasMoreElements()) {
-            URL url = resources.nextElement();
-            if (url.getProtocol().equals("jar")) {
-                JarURLConnection jar = (JarURLConnection) url.openConnection();
-                Enumeration<JarEntry> entries = jar.getJarFile().entries();
-                while (entries.hasMoreElements()) {
-                    JarEntry entry = entries.nextElement();
-                    if (entry.getName().startsWith(META_INF_RESOURCES)) {
-                        String sub = entry.getName().substring(META_INF_RESOURCES.length() + 1);
-                        if (!sub.isEmpty()) {
-                            if (entry.getName().endsWith("/")) {
-                                String dir = sub.substring(0, sub.length() - 1);
-                                knownDirectories.add(dir);
-                            } else {
-                                knownFiles.add(sub);
-                            }
-                        }
-                    }
-                }
-            }
-        }
-        substrateResourceBuildItemBuildProducer.produce(new SubstrateResourceBuildItem(
-                knownFiles.stream().map((s) -> META_INF_RESOURCES + "/" + s).collect(Collectors.toList())));
-        RuntimeValue<DeploymentInfo> deployment = template.createDeployment("test", knownFiles, knownDirectories,
+        RuntimeValue<DeploymentInfo> deployment = template.createDeployment("test", knownPaths.knownFiles,
+                knownPaths.knownDirectories,
                 launchMode.getLaunchMode(), shutdownContext);
 
         WebMetaData webMetaData = webMetadataBuildItem.getWebMetaData();

--- a/extensions/undertow/deployment/src/main/java/io/quarkus/undertow/deployment/UndertowStaticResourcesBuildStep.java
+++ b/extensions/undertow/deployment/src/main/java/io/quarkus/undertow/deployment/UndertowStaticResourcesBuildStep.java
@@ -1,0 +1,110 @@
+package io.quarkus.undertow.deployment;
+
+import java.net.JarURLConnection;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.jar.JarEntry;
+
+import io.quarkus.deployment.ApplicationArchive;
+import io.quarkus.deployment.annotations.BuildProducer;
+import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.builditem.ApplicationArchivesBuildItem;
+import io.quarkus.deployment.builditem.GeneratedResourceBuildItem;
+import io.quarkus.deployment.builditem.LaunchModeBuildItem;
+import io.quarkus.deployment.builditem.substrate.SubstrateResourceBuildItem;
+import io.quarkus.runtime.LaunchMode;
+
+public class UndertowStaticResourcesBuildStep {
+
+    protected static final String META_INF_RESOURCES_SLASH = "META-INF/resources/";
+    protected static final String META_INF_RESOURCES = "META-INF/resources";
+
+    @BuildStep
+    KnownPathsBuildItem scanStaticResources(ApplicationArchivesBuildItem applicationArchivesBuildItem,
+            BuildProducer<GeneratedResourceBuildItem> generatedResources,
+            List<GeneratedWebResourceBuildItem> generatedWebResources,
+            LaunchModeBuildItem launchModeBuildItem) throws Exception {
+        if (launchModeBuildItem.getLaunchMode() == LaunchMode.DEVELOPMENT) {
+            //we don't need to do this in development mode
+            //we serve directly from the project dir
+            return new KnownPathsBuildItem(Collections.emptySet(), Collections.emptySet());
+        }
+        //we need to check for web resources in order to get welcome files to work
+        //this kinda sucks
+        Set<String> knownFiles = new HashSet<>();
+        Set<String> knownDirectories = new HashSet<>();
+        for (ApplicationArchive i : applicationArchivesBuildItem.getAllApplicationArchives()) {
+            Path resource = i.getChildPath(META_INF_RESOURCES);
+            if (resource != null && Files.exists(resource)) {
+                Files.walk(resource).forEach(new Consumer<Path>() {
+                    @Override
+                    public void accept(Path path) {
+                        // Skip META-INF/resources entry
+                        if (resource.equals(path)) {
+                            return;
+                        }
+                        Path rel = resource.relativize(path);
+                        if (Files.isDirectory(rel)) {
+                            knownDirectories.add(rel.toString());
+                        } else {
+                            knownFiles.add(rel.toString());
+                        }
+                    }
+                });
+            }
+        }
+        Enumeration<URL> resources = getClass().getClassLoader().getResources(META_INF_RESOURCES);
+        while (resources.hasMoreElements()) {
+            URL url = resources.nextElement();
+            if (url.getProtocol().equals("jar")) {
+                JarURLConnection jar = (JarURLConnection) url.openConnection();
+                Enumeration<JarEntry> entries = jar.getJarFile().entries();
+                while (entries.hasMoreElements()) {
+                    JarEntry entry = entries.nextElement();
+                    if (entry.getName().startsWith(META_INF_RESOURCES_SLASH)) {
+                        String sub = entry.getName().substring(META_INF_RESOURCES_SLASH.length());
+                        if (!sub.isEmpty()) {
+                            if (entry.getName().endsWith("/")) {
+                                String dir = sub.substring(0, sub.length() - 1);
+                                knownDirectories.add(dir);
+                            } else {
+                                knownFiles.add(sub);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        for (GeneratedWebResourceBuildItem genResource : generatedWebResources) {
+            generatedResources.produce(new GeneratedResourceBuildItem(META_INF_RESOURCES_SLASH + genResource.getName(),
+                    genResource.getClassData()));
+            String sub = genResource.getName();
+            if (sub.startsWith("/")) {
+                sub = sub.substring(1);
+            }
+            if (!sub.isEmpty()) {
+                knownFiles.add(sub);
+                for (int i = 0; i < sub.length(); ++i) {
+                    if (sub.charAt(i) == '/') {
+                        knownDirectories.add(sub.substring(0, i));
+                    }
+                }
+            }
+        }
+        return new KnownPathsBuildItem(knownFiles, knownDirectories);
+    }
+
+    @BuildStep
+    void substrateResources(KnownPathsBuildItem paths, BuildProducer<SubstrateResourceBuildItem> substrate) {
+        for (String i : paths.knownFiles) {
+            substrate.produce(new SubstrateResourceBuildItem(META_INF_RESOURCES_SLASH + i));
+        }
+    }
+}

--- a/integration-tests/main/pom.xml
+++ b/integration-tests/main/pom.xml
@@ -57,6 +57,10 @@
         </dependency>
         <dependency>
             <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-swagger-ui</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
             <artifactId>quarkus-smallrye-opentracing</artifactId>
         </dependency>
         <dependency>

--- a/integration-tests/main/src/main/resources/application.properties
+++ b/integration-tests/main/src/main/resources/application.properties
@@ -39,3 +39,5 @@ quarkus.log.category.\"org.apache.kafka\".level=WARN
 quarkus.log.category.\"org.apache.zookeeper\".level=WARN
 
 microprofile.custom.value = 456
+
+quarkus.swagger-ui.always-include=true

--- a/integration-tests/main/src/test/java/io/quarkus/it/main/SwaggerUIITCase.java
+++ b/integration-tests/main/src/test/java/io/quarkus/it/main/SwaggerUIITCase.java
@@ -1,0 +1,24 @@
+/*
+ * Copyright 2018 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.quarkus.it.main;
+
+import io.quarkus.test.junit.SubstrateTest;
+
+@SubstrateTest
+public class SwaggerUIITCase extends SwaggerUITestCase {
+
+}

--- a/integration-tests/main/src/test/java/io/quarkus/it/main/SwaggerUITestCase.java
+++ b/integration-tests/main/src/test/java/io/quarkus/it/main/SwaggerUITestCase.java
@@ -1,0 +1,36 @@
+/*
+ * Copyright 2018 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.quarkus.it.main;
+
+import static org.hamcrest.Matchers.containsString;
+
+import org.junit.jupiter.api.Test;
+
+import io.quarkus.test.junit.QuarkusTest;
+import io.restassured.RestAssured;
+
+@QuarkusTest
+public class SwaggerUITestCase {
+
+    @Test
+    public void testSwaggerUi() {
+        RestAssured.when().get("/swagger-ui").then()
+                .body(containsString("#swagger-ui"))
+                .body(containsString("/openapi"));
+    }
+
+}

